### ISSUE TITLE
Merging kibana-ruby-auth

### ISSUE THREAD

**Issue by dav3860** [rashidkpc/kibana2#258]
Hi,

I open this issue to discuss what needs to be done to merge kibana-ruby-auth and kibana-ruby. I think that authentication and user management shouldn't be linked. For example, one could prefer to have LDAP auth but users and groups defined in Kibana. We could just separate auth_XXX.rb and create a users_XXX.rb for user management.
Are there any blocking bugs preventing kibana-ruby and -auth to be merged ?

/dav3860

**Comment by electrical** [comment]
Is there any reason to have authentication done in kibana?
If one runs kibana via passenger can't one use ldap in there then? ( just wondering )

**Comment by dav3860** [comment]
Yes, for role based access control for example. you can define groups and limit views to specific tags for example, letting sysadmins see the servers logs and network admins see the network equipment logs. Or to delegate a limited access to a helpdesk... Authentication could be done in LDAP too.
What in more, you have a nice welcome page (better than a pop-up) and you can use the backend to store favorites or search history.

**Comment by dav3860** [comment]
To me, it's a must-have for enterprise software.

**Comment by jamtur01** [comment]
Ditto +1.

**Comment by dav3860** [comment]
Most of the job is already done in kibana-ruby-auth. I think it lacks a few things but more people would help to fix them if there was only one branch.

**Comment by electrical** [comment]
okay makes sense now :-) :+1:

**Comment by heinemannj** [comment]
Have a look at request #240

For security reasons all back- and frontends should have authentication and role based access controls. 

Elasticsearch has no authentication and authorization by default - with the implementation of elasticsearch jetty plugin ( https://github.com/sonian/elasticsearch-jetty#readme ) elasticsearch can now handle SSL connections, support basic authentication and authorization, and log all or some incoming requests in plain text or json formats.

So we have to discuss two things:
a) Securing the elasticsearch frontends like kibana
100% agreement to above discussion ;-)

b) Securing the backendsystem elasticsearch with basic authentication over jetty for example
Therefor a possibility to define an elasticsearch UID and password in Kibana config file and further kibana code is mandantory.

**Comment by nicovpp** [comment]
Hi !

I am using passenger for running Kibana and already doing ldaps authentication on apache side.

I do have also tried kibana-ruby-auth and really like the user managment / tag restrictions / favorite search, but I think that there are some important missing function for people using ldap(s) : 
- ldaps support (certificate / ...)
- bindDN (or I maybe missed it)
- (I do agree that auth and user managment should be split)

Here is by example my apache configuration : 

```
    <location "/">
            AuthType Basic
            AuthBasicProvider ldap
            AuthzLDAPAuthoritative on
            AuthLDAPBindDN "CN=readonlyuser,ou=People,DC=myexample,DC=com"
            AuthLDAPBindPassword MySecretPassword:)
            !# --> need to deal with binddn / pw
            AuthLDAPURL "ldaps://W.X.Y.Z/ou=People,dc=myexample,dc=com?uid?sub?(objectClass=*)"
            !# --> need to deal with certificate and port 636
            AuthName "Kibana Restricted Access"
            Require valid-user
    </Location>
```

By the way, great job and would be great to merge :)

**Comment by prymitive** [comment]
+1 for splitting it, and maybe add an option to auto create user accounts in Kibana if they login correctly into LDAP

**Comment by dav3860** [comment]
@nicovpp : I have added a pull request for code doing an auth and user/group management separation. For example, I now use LDAP only for authentication against Active Directory, and all users and groups are provisioned directly on Kibana in elasticsearch. You have to provision your users in Kibana (shouldn't be too painful if you don't have too much Kibana users, and can be scripted). And groups are internal to Kibana. This works as expected and no specific binding is needed. It's OK for a basic but usable feature and it can be extended in the future.
For LDAPS and bindDN, NET::LDAP provides them, so it should be possible to implement.
@prymitive : yes, it's a good idea.

**Comment by nicovpp** [comment]
@dav3860 : Indeed it is not painful to provision users in kibana and specify ldap authentication for each one... Only a few users should be able to read logs. 
I wanted to put this version in my production environement but ldaps / bindDN are required... would be great if you implement these functions in the future.
I had a look but am not able to modify ldap_auth file (seems to deal with NET::LDAP for simple_tls encryption, and OpenSSL::SSL & OpenSSL::X509 for certificate & SSL connection)...

**Comment by electrical** [comment]
Where i will be using Kibana we are using Ldap for authentication and have allot of users.
Not sure if manual creation of users in kibana would be handy then.

**Comment by dav3860** [comment]
Separating identification, authentication and authorization doesn"t mean you can't use LDAP as a source for user authentication and identification/management. You would just have to specify LDAP as "auth_module" and "users_module". But the full user management in LDAP (with groups assignment, etc) isn't ready yet. For the moment, with kibana-ruby-auth, you can just authenticate against LDAP and manage your users internally.

**Comment by electrical** [comment]
Managing roles wouldn't be to bad to do in Kibana indeed.
Im looking forward to this functionality :-)

**Comment by dav3860** [comment]
You can already manage roles with kibana-ruby-auth. You can allow users or groups to only see messages with specific tags.

**Comment by electrical** [comment]
Cool :-) looking forward to have it all merged.
@rashidkpc How will functionality impact your new 'all js' version?

**Comment by heinemannj** [comment]
Any new status on merging kibana-ruby-auth and kibana-ruby?
What's about authentication and user management in Kibana3?

**Comment by dav3860** [comment]
In Kibana3, maybe we could use aliases. For each index, it's possible to create aliases with filters. So, it could be possible to create one for each user and each index, with a filter on specific types for example. Typically : "logstash-xx-xx-xx" would be the index, and "user1-xx-xx-xx" would be an alias for user1 limiting the view to one event type or more. Then with apache acting as a reverse proxy for ES, you could authenticate the user with LDAP for instance and rewrite all accesses to "/logstash-xx-xx-xx" to "/%{REMOTE_USER}-xx-xx-xx", thus transparently limiting the view to what the user is allowed to see.
A script using the ES API could be scheduled to create the aliases. Currently, there is no way to automatically create aliases with filters on index creation, even using index templates, but there is a pull request for this, I think. Or course, it would be easier if this was integrated to Kibana.
With this method, no "proxy" or authentication/authorization system would be needed except apache/nginx and ElasticSearch, I imagine.

**Comment by mrjcleaver** [comment]
+1 for needing this.

**Comment by diranged** [comment]
Just have to say that authentication on the frontend is important ... but so is the backend. I would love to put up an Apache or Nginx SSL reverse proxy in front of my ElasticSearch daemons and use HTTP AUTH (or SSL Certificate Verification) to make sure that my Kibana nodes have proper access to ElasticSearch directly, while no other clients do.
